File: uicorek/src/main/java/com/mozhimen/uicorek/bannerk/BannerK.java
package com.mozhimen.uicorek.bannerk;

import android.content.Context;
import android.content.res.TypedArray;
import android.util.AttributeSet;
import android.widget.FrameLayout;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.viewpager.widget.ViewPager;

import com.mozhimen.uicorek.R;
import com.mozhimen.uicorek.bannerk.commons.IBannerK;
import com.mozhimen.uicorek.bannerk.commons.IBannerKBindAdapter;
import com.mozhimen.uicorek.bannerk.commons.IBannerKIndicator;
import com.mozhimen.uicorek.bannerk.mos.BannerKMo;

import java.util.List;

/**
 * @ClassName BannerK
 * @Description TODO
 * @Author Kolin Zhao
 * @Date 2021/8/11 12:57
 * @Version 1.0
 */

/**
 * 核心问题:
 * 1. 如何实现UI的高度定制
 * 2. 作为有限的item如何实现无限轮播?
 * 3. Banner需要展示网络图片.如何将网络图片库和Banner组件进行解耦?
 * 4. 指示器样式各异, 如何实现指示器的高度定制?
 * 5. 如何设置Viewpager的滚动速度
 */
public class BannerK extends FrameLayout implements IBannerK {

    private BannerKDelegate delegate;

    public BannerK(@NonNull Context context) {
        this(context, null);
    }

    public BannerK(@NonNull Context context, @Nullable AttributeSet attrs) {
        this(context, attrs, 0);
    }

    public BannerK(@NonNull Context context, @Nullable AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        delegate = new BannerKDelegate(context, this);
        initCustomAttrs(context, attrs);
    }

    private void initCustomAttrs(Context context, AttributeSet attrs) {
        TypedArray typedArray = context.obtainStyledAttributes(attrs, R.styleable.BannerK);
        boolean autoPlay = typedArray.getBoolean(R.styleable.BannerK_bannerK_autoPlay, true);
        boolean loop = typedArray.getBoolean(R.styleable.BannerK_bannerK_loop, true);
        int intervalTime = typedArray.getInteger(R.styleable.BannerK_bannerK_intervalTime, -1);
        setAutoPlay(autoPlay);
        setLoop(loop);
        setIntervalTime(intervalTime);
        typedArray.recycle();
    }

    @Override
    public void setBannerData(int layoutResId, @NonNull List<? extends BannerKMo> models) {
        delegate.setBannerData(layoutResId, models);
    }

    @Override
    public void setBannerData(@NonNull List<? extends BannerKMo> models) {
        delegate.setBannerData(models);
    }

    @Override
    public void setBannerIndicator(IBannerKIndicator bannerKIndicator) {
        delegate.setBannerIndicator(bannerKIndicator);
    }

    @Override
    public void setAutoPlay(boolean autoPlay) {
        delegate.setAutoPlay(autoPlay);
    }

    @Override
    public void setLoop(boolean loop) {
        delegate.setLoop(loop);
    }

    @Override
    public void setIntervalTime(int intervalTime) {
        delegate.setIntervalTime(intervalTime);
    }

    @Override
    public void setCurrentItem(int position) {
        delegate.setCurrentItem(position);
    }

    @Override
    public void setBindAdapter(IBannerKBindAdapter bindAdapter) {
        delegate.setBindAdapter(bindAdapter);
    }

    @Override
    public void setOnPageChangeListener(ViewPager.OnPageChangeListener onPageChangeListener) {
        delegate.setOnPageChangeListener(onPageChangeListener);
    }

    @Override
    public void setOnBannerClickListener(OnBannerClickListener onBannerClickListener) {
        delegate.setOnBannerClickListener(onBannerClickListener);
    }

    @Override
    public void setScrollerDuration(int duration) {
        delegate.setScrollerDuration(duration);
    }
}

File: uicorek/src/main/java/com/mozhimen/uicorek/bannerk/BannerK.kt
package com.mozhimen.uicorek.bannerk

import android.content.Context
import android.util.AttributeSet
import androidx.viewpager.widget.ViewPager
import com.mozhimen.basick.basek.BaseKLayoutFrame
import com.mozhimen.uicorek.R
import com.mozhimen.uicorek.bannerk.commons.IBannerK
import com.mozhimen.uicorek.bannerk.commons.IBannerKBinder
import com.mozhimen.uicorek.bannerk.commons.IBannerKIndicator
import com.mozhimen.uicorek.bannerk.mos.MBannerKItem

/**
 * @ClassName BannerK
 * @Description 核心问题:
 * 1. 如何实现UI的高度定制
 * 2. 作为有限的item如何实现无限轮播?
 * 3. Banner需要展示网络图片.如何将网络图片库和Banner组件进行解耦?
 * 4. 指示器样式各异, 如何实现指示器的高度定制?
 * 5. 如何设置Viewpager的滚动速度
 * @Author mozhimen / Kolin Zhao
 * @Date 2022/4/19 23:26
 * @Version 1.0
 */
class BannerK @JvmOverloads constructor(context: Context, attrs: AttributeSet? = null, defStyleAttr: Int = 0) :
    BaseKLayoutFrame(context, attrs, defStyleAttr), IBannerK {
    private var _proxy: BannerKProxy = BannerKProxy(context, this)

    private var _autoPlay = true
    private var _loop = true
    private var _intervalTime = -1
    private var _scrollDuration = -1

    init {
        initAttrs(attrs, defStyleAttr)
    }

    override fun initAttrs(attrs: AttributeSet?, defStyleAttr: Int) {
        val typedArray = context.obtainStyledAttributes(attrs, R.styleable.BannerK)
        _autoPlay = typedArray.getBoolean(R.styleable.BannerK_bannerK_autoPlay, true)
        _loop = typedArray.getBoolean(R.styleable.BannerK_bannerK_loop, true)
        _intervalTime = typedArray.getInteger(R.styleable.BannerK_bannerK_intervalTime, -1)
        _scrollDuration = typedArray.getInteger(R.styleable.BannerK_bannerK_scrollDuration, -1)
        typedArray.recycle()

        setAutoPlay(_autoPlay)
        setLoop(_loop)
        setScrollDuration(_scrollDuration)
        setIntervalTime(_intervalTime)
    }

    /**
     * 设置Banner数据
     * @param mos List<BannerKMo>
     */
    override fun setBannerData(mos: List<MBannerKItem>) {
        _proxy.setBannerData(mos)
    }

    /**
     * 设置Banner数据
     * @param layoutResId Int
     * @param mos List<BannerKMo>
     */
    override fun setBannerData(layoutResId: Int, mos: List<MBannerKItem>) {
        _proxy.setBannerData(layoutResId, mos)
    }

    /**
     * 设置指示器
     * @param indicator IBannerKIndicator<*>
     */
    override fun setBannerIndicator(indicator: IBannerKIndicator<*>) {
        _proxy.setBannerIndicator(indicator)
    }

    /**
     * 设置自动轮播
     * @param autoPlay Boolean
     */
    override fun setAutoPlay(autoPlay: Boolean) {
        _proxy.setAutoPlay(autoPlay)
    }

    /**
     * 设置循环播放
     * @param loop Boolean
     */
    override fun setLoop(loop: Boolean) {
        _proxy.setLoop(loop)
    }

    /**
     * 设置滚动时长
     * @param duration Int page切换的时间长度 ,默认1000ms
     */
    override fun setScrollDuration(duration: Int) {
        _proxy.setScrollDuration(duration)
    }

    /**
     * 设置单页驻留时长
     * @param intervalTime Int 默认5000
     */
    override fun setIntervalTime(intervalTime: Int) {
        _proxy.setIntervalTime(intervalTime)
    }

    /**
     * 设置当前item
     * @param position Int
     */
    override fun setCurrentItem(position: Int) {
        _proxy.setCurrentItem(position)
    }

    /**
     * 设置绑定adapter
     * @param bindAdapter IBannerKBindAdapter
     */
    override fun setBindAdapter(bindAdapter: IBannerKBinder) {
        _proxy.setBindAdapter(bindAdapter)
    }

    /**
     * 设置页面改换监听器
     * @param onPageChangeListener OnPageChangeListener
     */
    override fun setOnPageChangeListener(onPageChangeListener: ViewPager.OnPageChangeListener) {
        _proxy.setOnPageChangeListener(onPageChangeListener)
    }

    /**
     * 设置页面点击监听器
     * @param IOnBannerClickListener OnBannerClickListener
     */
    override fun setOnBannerClickListener(IOnBannerClickListener: IBannerK.IOnBannerClickListener) {
        _proxy.setOnBannerClickListener(IOnBannerClickListener)
    }
}